### PR TITLE
[Inventory] Send top feature info

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

This branch encapsulate multiple PRs aiming at adding more information to the inventory payload.

### Describe how to test your changes

* Set `inventories_max_interval: 10` and `inventories_min_interval: 10` to trigger the firs payload after 10s (instead of the 5min default). With a fake HTTP endpoint check that you receive the inventory payload and its data are correct.
* since this PR also adds host metadata to the flare, create a flare (without uploading) and verify that it contains the metadata.
* we should also make sure the agent submits payloads to the new endpoint (visible at debug level in agent logs) successfully with 2xx responses, and confirm with metrics-intake that everything looks good on their end after staging/prod deploys

### Checklist

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

### CODE REVIEW

File: pkg/metadata/inventories/README.md [hunk 1]
```diff
@@ -36,19 +36,74 @@ The payload is a JSON dict with the following fields
...
 ## Example Payload
 
 Here an example of an inventory payload:
 
 ```
 {
-    "agent_metadata": {
-        "hostname_source": "os"
-    },
+   "agent_metadata": {
```
Comment (djmitche) [inline pkg/metadata/inventories/README.md:81]: The `cloud_provider` value is missing here.  It is optional, but might be best to have it appear anyway.

File: pkg/metadata/inventories/inventories.go [hunk 2]
```diff
@@ -54,22 +59,46 @@ var (
...
+	AgentConfigProxyHTTPS             AgentMetadataName = "config_proxy_https"
+	AgentInstallerVersion             AgentMetadataName = "install_method_installer_version"
+	AgentInstallTool                  AgentMetadataName = "install_method_tool"
+	AgentInstallToolVersion           AgentMetadataName = "install_method_tool_version"
+	AgentLogsTransport                AgentMetadataName = "logs_transport"
+	AgentSecurityEnabled              AgentMetadataName = "feature_cws_enabled"
+	AgentProcessEnabled               AgentMetadataName = "feature_process_enabled"
+	AgentNetworksEnabled              AgentMetadataName = "feature_networks_enabled"
+	AgentLogsEnabled                  AgentMetadataName = "feature_logs_enabled"
+	AgentComplianceEnabled            AgentMetadataName = "feature_cspm_enabled"
```
Comment (djmitche) [inline pkg/metadata/inventories/inventories.go:91]:
```suggestion
	AgentCWSEnabled              AgentMetadataName = "feature_cws_enabled"
	AgentProcessEnabled               AgentMetadataName = "feature_process_enabled"
	AgentNetworksEnabled              AgentMetadataName = "feature_networks_enabled"
	AgentLogsEnabled                  AgentMetadataName = "feature_logs_enabled"
	AgentCSPMEnabled            AgentMetadataName = "feature_cspm_enabled"
```

File: pkg/metadata/inventories/inventories.go [hunk 3]
```diff
@@ -54,22 +59,46 @@ var (
...
+	AgentConfigSite                   AgentMetadataName = "config_site"
+	AgentConfigLogsDDURL              AgentMetadataName = "config_logs_dd_url"
+	AgentConfigLogsSocks5ProxyAddress AgentMetadataName = "config_logs_socks5_proxy_address"
+	AgentConfigNoProxy                AgentMetadataName = "config_no_proxy"
+	AgentConfigProcessDDURL           AgentMetadataName = "config_process_dd_url"
+	AgentConfigProxyHTTP              AgentMetadataName = "config_proxy_http"
+	AgentConfigProxyHTTPS             AgentMetadataName = "config_proxy_https"
+	AgentInstallerVersion             AgentMetadataName = "install_method_installer_version"
+	AgentInstallTool                  AgentMetadataName = "install_method_tool"
+	AgentInstallToolVersion           AgentMetadataName = "install_method_tool_version"
```
Comment (djmitche) [inline pkg/metadata/inventories/inventories.go:85]:
```suggestion
	AgentInstallMethodInstallerVersion             AgentMetadataName = "install_method_installer_version"
	AgentInstallMethodTool                  AgentMetadataName = "install_method_tool"
	AgentInstallMethodToolVersion           AgentMetadataName = "install_method_tool_version"
```
Let's make sure the InitialCaps Go identifier and the snake_case key match.

File: pkg/metadata/inventories/inventories.go [hunk 4]
```diff
@@ -54,22 +59,46 @@ var (
...
-	// provider name in the agent metadata.
-	CloudProviderMetatadaName = "cloud_provider"
-	// HostnameSourceMetadataName is the field name to use to set the hostname
-	// source in the agent metadata.
-	HostnameSourceMetadataName = "hostname_source"
+	AgentCloudProvider                AgentMetadataName = "cloud_provider"
+	AgentHostnameSource               AgentMetadataName = "hostname_source"
+	AgentVersion                      AgentMetadataName = "agent_version"
+	AgentFlavor                       AgentMetadataName = "flavor"
+	AgentConfigApmDDURL               AgentMetadataName = "config_apm_dd_url"
```
Comment (djmitche) [inline pkg/metadata/inventories/inventories.go:74]:
```suggestion
	AgentConfigAPMDDURL               AgentMetadataName = "config_apm_dd_url"
```
(to match AgentAMPENabled, below)

**Review by djmitche — COMMENTED** [review]
There are really *three* lists of identifiers now (one in inventories.go and two in README.md).  I see that the JSON example is sorted.  Maybe we should sort all three?

**Comment by djmitche** [comment]
Per @olivielpeau ,
> I think we “only” need to change the endpoint to /api/v1/metadata and it should be ready to go.

so I'll try that :)

File: pkg/flare/metadata.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"path/filepath"
+
+	"github.com/DataDog/datadog-agent/pkg/metadata/inventories"
+	v5 "github.com/DataDog/datadog-agent/pkg/metadata/v5"
+	"github.com/DataDog/datadog-agent/pkg/util"
+)
+
+func addMetadata(tempDir, hostname, filename string, data []byte) error {
+	f := filepath.Join(tempDir, hostname, filename)
+	w, err := NewRedactingWriter(f, os.ModePerm, true)
```
Comment (djmitche) [inline pkg/flare/metadata.go:21]: This risks causing cause failures on `main` if not merged carefully with https://github.com/DataDog/datadog-agent/pull/9682.

File: pkg/forwarder/telemetry.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ var (
 	v1IntakeEndpoint       = transaction.Endpoint{Route: "/intake/", Name: "intake"}
 	v1SketchSeriesEndpoint = transaction.Endpoint{Route: "/api/v1/sketches", Name: "sketches_v1"} // nolint unused for now
 	v1ValidateEndpoint     = transaction.Endpoint{Route: "/api/v1/validate", Name: "validate_v1"}
+	v1MetadataEndpoint     = transaction.Endpoint{Route: "/api/v1/metadata", Name: "metadata_v1"}
```
Comment (djmitche) [inline pkg/forwarder/telemetry.go:25]: I'm a little unclear why there's a "new" v1 endpoint when we already have a v2 metadata endpoint?

File: pkg/forwarder/forwarder.go [hunk 7]
```diff
@@ -464,7 +464,8 @@ func (f *DefaultForwarder) SubmitAgentChecksMetadata(payload Payloads, extra htt
 
 // SubmitMetadata will send a metadata type payload to Datadog backend.
 func (f *DefaultForwarder) SubmitMetadata(payload Payloads, extra http.Header) error {
-	return f.submitV1IntakeWithTransactionsFactory(payload, extra, f.createHTTPTransactions)
+	transactions := f.createHTTPTransactions(v1MetadataEndpoint, payload, false, extra)
+	return f.sendHTTPTransactions(transactions)
```
Comment (djmitche) [inline pkg/forwarder/forwarder.go:468]: Is this correct, or do we want a config flag to allow these to still be sent to `/intake` by default?

**Review by djmitche — APPROVED** [review]

File: pkg/forwarder/telemetry.go (hunk 6 above)
Comment (olivielpeau) [reply pkg/forwarder/telemetry.go:25]: The v2 metadata endpoint doesn't actually exist (the "v2 metadata code" the Agent was added a long time ago, but should be cleaned up as the endpoint never actually existed)

File: pkg/forwarder/forwarder.go (hunk 7 above)
Comment (olivielpeau) [reply pkg/forwarder/forwarder.go:468]: * metadata payloads other than the inventories payload should still go to `/intake` (but I haven't checked what payloads are sent with `SubmitMetadata`)
* I think we're fine without a flag for the inventories payload

Comment (djmitche) [reply pkg/forwarder/forwarder.go:468]: I guess the "metadata payloads other than inventories" are host metadata and agent checks metadata?  Those are still using `/intake` as desired.

File: pkg/forwarder/telemetry.go (hunk 6 above)
Comment (djmitche) [reply pkg/forwarder/telemetry.go:25]: OK, I'll make a PR to remove that.

File: pkg/flare/metadata.go (hunk 5 above)
Comment (djmitche) [reply pkg/flare/metadata.go]: #9682 has landed, and I merged from master.  I caught this issue in the merge.

File: pkg/forwarder/forwarder.go (hunk 7 above)
Comment (olivielpeau) [reply pkg/forwarder/forwarder.go:493]: Correct, and yes these payloads should continue to be sent to `/intake`

**Review by olivielpeau — APPROVED** [review]
Let's add the following to the QA plan:
1. since this PR also adds host metadata to the flare, let's make sure we cover the flare pieces in the QA plan
2. we should also make sure the agent submits payloads to the new endpoint (visible at debug level in agent logs) successfully with 2xx responses, and confirm with metrics-intake that everything looks good on their end after staging/prod deploys

One important comment on the HTTP header, other than that looks good to me!

File: pkg/forwarder/forwarder.go [hunk 8]
```diff
@@ -489,7 +489,8 @@ func (f *DefaultForwarder) SubmitAgentChecksMetadata(payload Payloads, extra htt
 
 // SubmitMetadata will send a metadata type payload to Datadog backend.
 func (f *DefaultForwarder) SubmitMetadata(payload Payloads, extra http.Header) error {
-	return f.submitV1IntakeWithTransactionsFactory(payload, extra, f.createHTTPTransactions)
+	transactions := f.createHTTPTransactions(endpoints.V1MetadataEndpoint, payload, false, extra)
```
[the remark this answers is not in the document]
Comment (djmitche) [reply pkg/forwarder/forwarder.go:492]: I can add that to extra, but I'm basing that only on your suggesting that I do so -- why would it be required?

[the remark this answers is not in the document]
Comment (truthbk) [reply pkg/forwarder/forwarder.go:492]: I think we should add it, not sure it's required, but the metadata payload is a json payload, so it feels like the right thing to do even if the backend doesn't care.

**Review by truthbk — APPROVED** [review]

**Comment by djmitche** [comment]
It looks like all of the CI jobs that might be problematic here have completed.  I think the kitchen's a little backed up.

**Pull request merged**